packages/cinterop/src/jvm/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt

```diff
@@ -1443,6 +1414,113 @@ actual object RealmInterop {
     }
 }
 
+/**
+ * A factory and container for various resources that can be freed when calling [free].
+ *
+ * @see memScope
+ */
+private class MemScope {
+    val values: MutableSet<realm_value_t> = mutableSetOf()
+
+    fun managedRealmValue(value: RealmValue): realm_value_t {
+        val element = capiRealmValue(value)
+        values.add(element)
+        return element
+    }
+
+    fun free() {
+        values.map { realmc.realm_value_t_cleanup(it) }
+    }
+}
+
+/**
+ * A scope providing a [MemScope] that collects the created resources and automatically frees them
+ * upon exiting the scope.
+ */
+private fun <R> memScope(block: MemScope.() -> R): R {
+    val scope = MemScope()
+    val result: R = block(scope)
+    scope.free()
```

Comment:
Shouldn't this be in a `try/finally` block so we always free data even if there is an exception?